tools: add WPT updater for specific subsystems

mertcanaltin opened the pull request:
a WPT updater that specifically targets the url subsystem for more controlled and focused updates.

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+  schedule:
+    - cron: 0 0 * * *
+
+permissions:
+  contents: write
+  pull-requests: write
```
```suggestion
permissions: {}
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+
+jobs:
+  update-wpt:
+    name: Update WPT
```
Putting permissions here is safer IMO

mertcanaltin replied:
thanks

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633  # v4.1.2
```
```suggestion
        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332  # v4.1.7
        with:
          persist-credentials: false
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+
+          rm -rf test/fixtures/tmp-wpt
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@c5a7806660adbe173f04e3e038b0ccdcd758773c  # v6.1.0
```
Can you use create or update pull request, like in the `tools.yml` file?

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+        run: |
+          rm -rf test/fixtures/wpt && mkdir test/fixtures/wpt &&
+
+          git clone https://github.com/web-platform-tests/wpt.git --depth=1 test/fixtures/tmp-wpt &&
```
You could replace this with an extra call to `actions/checkout`?

And sparse checkout only the folders we need directly into the folder we want them to go?

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+        uses: peter-evans/create-pull-request@c5a7806660adbe173f04e3e038b0ccdcd758773c  # v6.1.0
+        with:
+          base: main
+          branch: wpt-update
```
```suggestion
          branch: actions/wpt-update
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+          branch: wpt-update
+          title: Update WPT
+          body: Automated update of the WPT
+          commit-message: 'chore: update WPT'
```
I don't think `chore` is a valid subsystem

mertcanaltin replied:
I updated it to ‘test: update wpt test’

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,46 @@
+on:
+  workflow_dispatch:
+  schedule:
+    - cron: 0 0 * * *
```
A comment saying how often this is would be nice

mertcanaltin replied:
thanks

mertcanaltin commented:
thanks for your suggestions

mertcanaltin commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,74 @@
+          mv test/fixtures/tmp-wpt/xhr test/fixtures/wpt/xhr &&
+          rm -rf test/fixtures/tmp-wpt
+      - name: Create or Update Pull Request
+        uses: peter-evans/create-pull-request@c5a7806660adbe173f04e3e038b0ccdcd758773c  # v6.1.0
```
Oh, I forgot I was fixing this

mertcanaltin commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,74 @@
+      - name: Update WPT
+        run: |
+          rm -rf test/fixtures/wpt && mkdir test/fixtures/wpt &&
+          - name: Clone WPT Repository
```
Are you talking about the lint error here?

RedYetiDev replied:
I mean that the checkout is indented to much, so it's part of the run statement

RedYetiDev replied:
I think it's resolved in the latest commit sorry for the confusion

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,52 @@
+        with:
+            repository: web-platform-tests/wpt
+            ref: ${{ github.sha }}
+            path: test/fixtures/tmp-wpt
```
```suggestion
            path: test/fixtures/wpt
```

mertcanaltin replied:
thanks, ı used

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,52 @@
+          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/
+      - name: Move WPT Files
+        run: |
+            mv test/fixtures/tmp-wpt/LICENSE.md test/fixtures/wpt/LICENSE.md &&
+            mv test/fixtures/tmp-wpt/common test/fixtures/wpt/common &&
+            mv test/fixtures/tmp-wpt/eventsource test/fixtures/wpt/eventsource &&
+            mv test/fixtures/tmp-wpt/fetch test/fixtures/wpt/fetch &&
+            mv test/fixtures/tmp-wpt/interfaces test/fixtures/wpt/interfaces &&
+            mv test/fixtures/tmp-wpt/mimesniff test/fixtures/wpt/mimesniff &&
+            mv test/fixtures/tmp-wpt/resources test/fixtures/wpt/resources &&
+            mv test/fixtures/tmp-wpt/service-workers test/fixtures/wpt/service-workers &&
+            mv test/fixtures/tmp-wpt/storage test/fixtures/wpt/storage &&
+            mv test/fixtures/tmp-wpt/websockets test/fixtures/wpt/websockets &&
+            mv test/fixtures/tmp-wpt/xhr test/fixtures/wpt/xhr &&
+            rm -rf test/fixtures/tmp-wpt/.git
```
```suggestion
            rm -rf test/fixtures/wpt/.git
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,52 @@
+          repository: web-platform-tests/wpt
+          ref: ${{ github.sha }}
+          path: test/fixtures/tmp-wpt
+          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/
```
```suggestion
          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/ LICENSE.md
```

Also this can go on separate lines, right?

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+          repository: web-platform-tests/wpt
+          ref: ${{ github.sha }}
+          path: test/fixtures/wpt
+          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/ LICENSE.md
```
```suggestion
          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/ LICENSE.md
          persist-credentials: false
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+      - name: Move WPT Files
+        run: |
+            rm -rf test/fixtures/wpt/.git
+      - name: Create Pull Request
+        uses: peter-evans/create-pull-request@c5a7806660adbe173f04e3e038b0ccdcd758773c  # v6.1.0
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+          branch: actions/wpt-update
+          title: Update WPT
+          body: Automated update of the WPT
+          commit-message: 'test: update wpt test'
```
```suggestion
      - name: Create Or Update Pull Request
        uses: gr2m/create-or-update-pull-request-action@86ec1766034c8173518f61d2075cc2a173fb8c97  # v1.9.4
        env:
          GITHUB_TOKEN: ${{ secrets.GH_USER_TOKEN }}
        with:
          author: Node.js GitHub Bot <github-bot@iojs.org>
          body: This is an automated update of the WPT to ${{ env.NEW_VERSION }}.
          branch: actions/update-wpt  # Custom branch *just* for this Action.
          commit-message: 'test: update WPT to ${{ env.NEW_VERSION }}'
          labels: test
          title: 'test: update WPT to ${{ env.NEW_VERSION }}'
          update-pull-request-title-and-body: true
```

mertcanaltin replied:
Create Or Update Pull Request: Shouldn't the recommendation only cover Create Pull Request?

RedYetiDev replied:
Create or update pull request will allow it to update the WPT files again and again

mertcanaltin replied:
thanks

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+          sparse_checkout: common/ eventsource/ fetch/ interfaces/ mimesniff/ resources/ service-workers/ storage/ websockets/ xhr/ LICENSE.md
+      - name: Move WPT Files
+        run: |
+            rm -rf test/fixtures/wpt/.git
```
```suggestion
            rm -rf test/fixtures/wpt/.git
            echo "NEW_VERSION=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
```

RedYetiDev reviewed (commented):
I also feel like this could be moved to a `dep_updater` file.

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+        uses: actions/checkout@v2
+        with:
+          repository: web-platform-tests/wpt
+          ref: ${{ github.sha }}
```
```suggestion
```

RedYetiDev replied:
```suggestion
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+      pull-requests: write
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332  # v4.1.7
+        with:
+          persist-credentials: false
+      - name: Update WPT
+        run: |
+          rm -rf test/fixtures/wpt
+          mkdir test/fixtures/wpt
```
```suggestion
```

RedYetiDev replied:
```suggestion
      - name: Checkout Repository
        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332  # v4.1.7
        with:
          persist-credentials: false
          sparse-checkout: |
            !test/fixtures/wpt
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,42 @@
+          rm -rf test/fixtures/wpt
+          mkdir test/fixtures/wpt
+      - name: Clone WPT Repository
+        uses: actions/checkout@v2
```
```suggestion
        uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # v4.1.7
```

RedYetiDev commented:
By the way, I'm trying to update the WPT files in #54468.

IMO this could be done via 
```
git node wpt <...>
```

Because the README.md file also needs to be updated.

mertcanaltin commented:
> By the way, I'm trying to update the WPT files in #54468.
> 
> IMO this could be done via
> 
> ```
> git node wpt <...>
> ```
> 
> Because the README.md file also needs to be updated.

Do you have a suggestion, would that be enough for this?

```
- Name: Update README.md file
        Run: |
            # Update the README.md file with the new WPT version
            # Assuming you have a script or command to do this
            ./scripts/update-readme.sh ‘${{{ env.NEW_VERSION }}’

      - Name: Commit to Changes
        Run: |
            git config --global user.name ‘Node.js GitHub Bot’
            git config --global user.email ‘github-bot@iojs.org’
            git add test/fixtures/wpt README.md
            git commit -m ‘test: Update WPT to ${{{ env.NEW_VERSION }} and update README.md’
```

RedYetiDev commented:
In my opinion, the entire WPT update process could be moved to a dep-updater script, using `git node` for each section. But, there are a lot of cases where updating these tests cause problems (just look at the errors in my PR), so maybe keeping it manual is good?

I'm not sure.

panva reviewed (dismissed):
In general I don't think updating WPTs automatically is a good idea. I can't find the issue now but I believe we've already had that discussion in the past, that's not to say we can't have it again... So:

- The way to update WPTs is through the `git node wpt ...` command from [node-core-utils](https://github.com/nodejs/node-core-utils), not checking out the WPT repo and copying over files
- WPTs have dependencies in `resources`, `interfaces`, or generally a new folder that is not part of the checkout and requires manual intervention
- Pulling in new WPTs often makes new WPTs fail, that also requires manual intervention
- There are tentative WPTs that are not tied to a finished spec change, that will fail too
- If there was an automated process then WPTs should be updated by their subsystem, not all at once, but that again raises an issue when that subsystem WPT change requires to pull in a file from a different directory that also needs to be updated
- Some WPT subsystems currently cannot be updated (e.g. WebCryptoAPI) because we don't have a way to conditionally mark tests as failing depending on the system they run on, in WebCryptoAPI's case - #54127, https://github.com/nodejs/node/pull/52222#issuecomment-2021333019, RHEL's linked OpenSSL makes some vectors pass that don't pass anywhere else.

In my opinion a sensible way to update WPTs would be to pull a WPT subsystem from upstream (using `git node wpt ...`), run it, if it passes - open a PR, if it fails - have a keepalive issue where pending WPT updates are tracked.

FWIW this could also be hooked into the daily wpt job that runs the most up to date WPT checkout for the purpose of [uploading the current state to wpt.io](https://wpt.fyi/results/?label=master&label=experimental&product=node.js&product=deno&product=chrome&product=edge&product=firefox&product=safari&product=flow&product=ladybird&view=subtest&q=node.js%3A%21missing).

Remember that WPTs may very well fail for a long time when they are a) tentative or b) require a semver-major change to pass.

panva commented:
cc @nodejs/web-standards

legendecas commented:
yeah, we had discussion at https://github.com/nodejs/node/pull/50567.

mertcanaltin commented:
I'm closing this place, thank you very much for the comments

panva commented:
@anonrig That's pretty much a version of what I think we should be doing, only scoped to a particular WPT suite.

> In my opinion a sensible way to update WPTs would be to pull a WPT subsystem from upstream (using git node wpt ...), run it, if it passes - open a PR, if it fails - have a keepalive issue where pending WPT updates are tracked.

anonrig commented:
Perfect @mertcanaltin seems interested in pursuing this. Mertcan can you add a WPT updater but only for `url`? Make sure to have a list  for the scopes, so we can add them depending on people's interest.

mertcanaltin commented:
This looks great, I'm gonna do it

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,82 @@
+          persist-credentials: false
+
+      # Generate bot token
+      - uses: actions/create-github-app-token@v1
+        id: app-token
+        with:
+          app-id: ${{ vars.APP_ID }}
+          private-key: ${{ secrets.APP_PEM }}
+
+      - name: Retrieve GitHub App User ID
+        id: get-user-id
+        run: echo "user-id=$(gh api "/users/${{ steps.app-token.outputs.app-slug }}[bot]" --jq .id)" >> "$GITHUB_OUTPUT"
+        env:
+          GH_TOKEN: ${{ steps.app-token.outputs.token }}
```
Can't this just be `GH_USER_TOKEN`?

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,82 @@
+      - name: Handle test failures and success
+        run: |
+          if [[ "$(git status --porcelain)" != "" ]]; then
+            git config --global user.name "${{steps.app-token.outputs.app-slug}}[bot]"
+            git config --global user.email "${{steps.get-user-id.outputs.user-id}}+${{ steps.app-token.outputs.app-slug }}[bot]@users.noreply.github.com"
```
Ditto

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,82 @@
+          if [[ "$(git status --porcelain)" != "" ]]; then
+            git config --global user.name "${{steps.app-token.outputs.app-slug}}[bot]"
+            git config --global user.email "${{steps.get-user-id.outputs.user-id}}+${{ steps.app-token.outputs.app-slug }}[bot]@users.noreply.github.com"
+            git checkout -b update-wpt-${{ matrix.subsystem }}
```
```suggestion
            git checkout -b actions/update-wpt-${{ matrix.subsystem }}
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,82 @@
+        env:
+          GH_TOKEN: ${{ steps.app-token.outputs.token }}
+
+      - name: Open a PR for the subsystem update
+        if: steps.push-changes.outputs.createPR
+        continue-on-error: true
+        run: |
+          gh pr create --head update-wpt-${{ matrix.subsystem }} \
+          --title "test: update WPT for ${{ matrix.subsystem }} to ${{ env.NEW_VERSION }}" \
+          --body "This is an automated update of the WPT for ${{ matrix.subsystem }} to ${{ env.NEW_VERSION }}."
+        env:
+          GH_TOKEN: ${{ steps.app-token.outputs.token }}
```
This should be `create-or-update-pr`, right? Like in the rest of the updators.

mertcanaltin replied:
thanks

RedYetiDev reviewed (commented):
IMO this should align more with <https://github.com/nodejs/node/blob/main/.github/workflows/update-openssl.yml>, and the other updaters.

---

Also, could this go in `dep-updaters`?

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,82 @@
+          - websockets
+
+    steps:
+      - uses: actions/checkout@v4
```
For security reasons, I believe all of these should be referenced by hash, not tag.

mertcanaltin replied:
applied, thanks

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,67 @@
+      - uses: actions/checkout@v4@<hash>
+        with:
+          persist-credentials: false
+
+      - name: Set up GitHub Authentication
+        run: echo "GitHub Authentication set up"
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,67 @@
+          - url
+
+    steps:
+      - uses: actions/checkout@v4@<hash>
```
Can you add the hash?

mertcanaltin replied:
Do you have any idea how I can get this hash?

marco-ippolito replied:
check the latest v4 tag, you get the commit hash, and add a comment with the tag

RedYetiDev replied:
Or check any of the other workflows in use

mertcanaltin replied:
i added, thanks

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,67 @@
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Open a PR for the subsystem update
+        if: steps.push-changes.outputs.createPR
+        continue-on-error: true
+        run: |
+          gh pr create --head actions/update-wpt-${{ matrix.subsystem }} \
+          --title "test: update WPT for ${{ matrix.subsystem }} to ${{ env.NEW_VERSION }}" \
+          --body "This is an automated update of the WPT for ${{ matrix.subsystem }} to ${{ env.NEW_VERSION }}."
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
WDYT about the create-or-update-pr action?

mertcanaltin replied:
I think it is healthier to have an update

mertcanaltin replied:
I thought of this to indicate the purpose of the mr that will occur here, since this is an update, wouldn't it be good to update it, thank you for your comments, I will apply it, I'm just curious 🙏❤️🚀

mertcanaltin replied:
I applied here  🙏

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,62 @@
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Run WPT tests
+        run: |
+          git node wpt run --subsystem=${{ matrix.subsystem }}
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
      - name: Run WPT tests
        run: |
          git node wpt run --subsystem=${{ matrix.subsystem }}
        env:
          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

Maybe if they fail, continue with opening the PR, but add the output to the PR body letting maintainers know that the WPT tests failed?

mertcanaltin replied:
thank you I sent the commit for it

mertcanaltin commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,62 @@
+        run: |
+          git node wpt update --subsystem=${{ matrix.subsystem }}
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
thanks 🙏

marco-ippolito commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,70 @@
+          - url
+
+    steps:
+      - uses: actions/checkout@v4@6d193bf28034eafb982f37bd894289fe649468fc
```
```suggestion
      - uses: actions/checkout@6d193bf28034eafb982f37bd894289fe649468fc
```
Can you add a comment with the tag that this commit points to?

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,70 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        subsystem:
```
Can you make the subsystem be an argument that's able to be passed to workflow_dispatch? That way we can trigger this for specific subsystem(s) that aren't necessarily run in a cron as needed too.

https://github.com/nodejs/node/blob/4f1fe8a01553ecfe8e63a484165313fe3fd8c10e/.github/workflows/daily-wpt-fyi.yml#L9-L13

https://github.com/nodejs/node/blob/4f1fe8a01553ecfe8e63a484165313fe3fd8c10e/.github/workflows/daily-wpt-fyi.yml#L29-L32

mertcanaltin replied:
applied thank you

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,70 @@
+jobs:
+  wpt-subsystem-update:
+    runs-on: ubuntu-latest
+    strategy:
```
this needs `fail-fast: false`

mertcanaltin replied:
thanks

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,70 @@
+
+      - name: Update WPT for subsystem ${{ matrix.subsystem }}
+        run: |
+          git node wpt update --subsystem=${{ matrix.subsystem }}
```
Where is this git extension coming from? The one we have is 

```
git node wpt ${{ matrix.subsystem }}
```

And FWIW I don't think this action install `@node-core/utils` globally to have that available.

mertcanaltin replied:
ı applied your suggestion : `git node wpt ${{ matrix.subsystem }}`

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,70 @@
+      - name: Run WPT tests
+        id: run-wpt-tests
+        run: |
+          git node wpt run --subsystem=${{ matrix.subsystem }} || echo "WPT tests failed" > wpt-failure.txt
```
I can't figure out where `git node wpt run --subsystem` is coming from either.

I think running the WPTs in the action has become redundant since this will open a PR regardless of the outcome. That PR will run github CI straight away and will need full CI afterwards too.

mertcanaltin replied:
I removed the part that runs the WPT Tests as you said

panva reviewed (dismissed):
Since this will open a PR regardless please remove running WPT portion bit.

Also, some suggestions to improve the workflow_dispatch usability with flexible subsystem input.

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,60 @@
+      subsystem:
+        description: Subsystem to run the update for
+        required: false
+        default: url
```
This doesn't give way to have it run for multiple subsystems at a later point, please see the daily wpt fyi runner.

mertcanaltin replied:
can ı use it `default: '["current", "lts/*", "lts/-1"]'`

panva replied:
This is tied to how you construct the matrix later. As is it won't work.

mertcanaltin replied:
I think I need to make it compatible with the url

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,60 @@
+    - cron: 0 0 * * *
+  workflow_dispatch:
+    inputs:
+      subsystem:
+        description: Subsystem to run the update for
+        required: false
+        default: url
```
```suggestion
        default: '["url"]'
```

panva replied:
```suggestion
      subsystems:
        description: Subsystem to run the update for
        required: false
        default: '["url"]'
```

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,60 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        subsystem: [ '${{ github.event.inputs.subsystem || "url" }}' ]
```
```suggestion
        subsystem: ${{ fromJSON(github.event.inputs.subsystems || '["url"]') }}
```

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,60 @@
+
+      - name: Update WPT for subsystem ${{ matrix.subsystem }}
+        run: |
+          git node wpt ${{ matrix.subsystem }}
```
this will not work until node-core-utils is [installed](https://github.com/nodejs/node-core-utils#install).

panva commented:
Will you test this in https://github.com/nodejs/node-auto-test?

mertcanaltin commented:
https://github.com/mertcanaltin/node-auto-test/actions/runs/10972255885/job/30468306292
I did a test run here and it seems to give an error to the github token

panva commented:
> [mertcanaltin/node-auto-test/actions/runs/10972255885/job/30468306292](https://github.com/mertcanaltin/node-auto-test/actions/runs/10972255885/job/30468306292) I did a test run here and it seems to give an error to the github token

I think it seems to suggest the script is ending prematurely. i.e. not syntactically correct.

mertcanaltin commented:
Passed 🎉  https://github.com/mertcanaltin/node-auto-test/actions/runs/10972357345/job/30468531837

The issue was that the if block wasn’t properly closed. I just added the missing fi at the end, and now everything works smoothly.

panva commented:
> Passed 🎉 [mertcanaltin/node-auto-test/actions/runs/10972357345/job/30468531837](https://github.com/mertcanaltin/node-auto-test/actions/runs/10972357345/job/30468531837)
> 
> The issue was that the if block wasn’t properly closed. I just added the missing fi at the end, and now everything works smoothly.

Can you run dispatch with two subsystems? url and WebCryptoAPI. The latter should open a PR.

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,65 @@
+  contents: read
+
+jobs:
+  wpt-subsystem-update:
```
```suggestion
  wpt-subsystem-update:
    if: github.repository == 'nodejs/node' || github.event_name == 'workflow_dispatch'
```

mertcanaltin commented:
https://github.com/mertcanaltin/node-auto-test/actions/runs/10972538059 I ran for both of them

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,66 @@
+            echo "No changes to commit."
+          fi
+        env:
+          GH_USER_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
          GH_TOKEN: ${{ secrets.GH_USER_TOKEN }}
```

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,66 @@
+        if: steps.push-changes.outputs.createPR
+        uses: peter-evans/create-pull-request@v4
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
```
Ditto

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,66 @@
+
+      - name: Open or update PR for the subsystem update
+        if: steps.push-changes.outputs.createPR
+        uses: peter-evans/create-pull-request@v4
```
You should use create-or-update-pr

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,66 @@
+            git config --global user.email "${{ github.actor }}@users.noreply.github.com"
+            git checkout -b actions/update-wpt-${{ matrix.subsystem }}
+            git add .
+            git commit -m "test: update WPT for ${{ matrix.subsystem }} to ${{ env.NEW_VERSION }}"
```
Where is NEW_VERSION being calculated?

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,66 @@
+        with:
+          persist-credentials: false
+
+      - name: Install node-core-utils
+        run: |
+          npm install -g node-core-utils
```
```suggestion
      - name: Install @node-core/utils
        run: |
          npm install -g @node-core/utils
```

panva commented:
https://github.com/mertcanaltin/node-auto-test/actions/runs/10972538059/job/30468925665#step:4:7

It would appear `@node-core/utils` needs its config file with credentials to run this. Maybe an empty config file would do? I dunno, I'm short on time to look into this.

RedYetiDev commented:
You could also move all this logic into a dependency updater script, and call it like the rest of the updaters.

mertcanaltin commented:
> You could also move all this logic into a dependency updater script, and call it like the rest of the updaters.

ı applied the latest updates, thank you

mertcanaltin commented:
> You could also move all this logic into a dependency updater script, and call it like the rest of the updaters.

node/.github/workflows/update-openssl.yml

I looked here as an example

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,76 @@
+        run: |
+          # Hypothetically, you can get the latest version from a specific file or git log
+          new_version=$(git log -1 --pretty=format:"%h")
+          echo "new_version=$new_version" >> $GITHUB_ENV
```
You can only obtain the new version after `git node wpt` is done. And you can obtain the new version with `jq` extraction using the `test/fixtures/wpt/versions.json` file as input

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,75 @@
+
+      - name: Calculate new version for WPT using jq
+        run: |
+          new_version=$(jq -r '.url.commit' test/fixtures/wpt/versions.json)
```
```suggestion
          new_version=$(jq -r '.${{ matrix.subsystem }}.commit' test/fixtures/wpt/versions.json)
```

panva commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,75 @@
+
+      - name: Calculate new version for WPT using jq
+        run: |
+          new_version=$(jq -r '.${{ matrix.subsystem }}.commit' test/fixtures/wpt/versions.json)
```
```suggestion
          new_version=$(jq -r '.["${{ matrix.subsystem }}"].commit' test/fixtures/wpt/versions.json)
```

panva commented:
Look at .github/workflows/auto-start-ci.yml for `Setup @node-core/utils`

RedYetiDev commented on .github/workflows/update-wpt.yml:
```diff
@@ -0,0 +1,75 @@
+        run: |
+          npm install -g @node-core/utils
+
+      - name: Calculate new version for WPT using jq
+        run: |
+          new_version=$(jq -r '.["${{ matrix.subsystem }}"].commit' test/fixtures/wpt/versions.json)
+          echo "new_version=$new_version" >> $GITHUB_ENV
```
This should go after `git node wpt`

panva commented:
I believe the last 2 points weren't adressed. ncu is not configured and the new version variable is used before it is set

mertcanaltin commented:
> I believe the last 2 points weren't addressed. ncu is not configured and the new version variable is used before it is set

Thanks a lot for your detailed feedback! I’ve now addressed the last two points you mentioned:

I've added a step to configure ncu (node-core-utils) with the necessary GitHub token.
I moved the calculation of the new_version earlier in the workflow to ensure it's set before being referenced.

panva commented:
@nodejs/actions please review

RedYetiDev commented:
FYI I was looking at PRs and I saw #50567 cc @legendecas
